src/utils/concurrency.ts

```diff
@@ -0,0 +1,35 @@
+import PQueue, { QueueAddOptions } from 'p-queue';
+
+type Limiter = () => Promise<{ remaining: number }>;
+type Task<T> = () => Promise<T>;
+
+export class ConcurrencyLimiter {
+  private readonly queue = new PQueue({ concurrency: 0 });
+  private watcher?: NodeJS.Timeout;
+
+  constructor(private readonly limiter: Limiter) {
+    const updateLimit = async () => {
+      try {
+        const { remaining } = await limiter();
+        this.queue.concurrency = remaining;
```

Comment:
Not safe to mutate such property. Behavior is not documented.

Reply:
It is a public setter, as such it should be safe to use with understanding that concurrency won't be enforced at all times. Anyway this will disappear after fixing the comment below.